local-storage/src/provider.rs

```diff
@@ -0,0 +1,383 @@
+        self.conn.execute(
+            "CREATE TABLE IF NOT EXISTS blocks (
+                id INTEGER PRIMARY KEY,
+                cid TEXT NOT NULL,
+                data BLOB
+            )",
+            (),
+        )?;
+
+        // Create links table
+        self.conn.execute(
+            "CREATE TABLE IF NOT EXISTS links(
+                id INTEGER PRIMARY KEY,
+                root_cid TEXT,
+                block_cid TEXT NOT NULL,
+                block_id INTEGER,
+                FOREIGN KEY (block_id) REFERENCES blocks (id)
+            )",
+            (),
+        )?;
+
+        // Create indices
+        self.conn.execute(
+            "CREATE UNIQUE INDEX IF NOT EXISTS blocks_cid on blocks(cid)",
+            (),
+        )?;
+        self.conn.execute(
+            "CREATE INDEX IF NOT EXISTS links_root_cid on links(root_cid)",
+            (),
+        )?;
+        self.conn.execute(
+            "CREATE UNIQUE INDEX IF NOT EXISTS links_block_cid on links(block_cid)",
+            (),
+        )?;
+
+        Ok(())
+    }
+}
+
+impl StorageProvider for SqliteStorageProvider {
+    fn import_block(&self, block: &StoredBlock) -> Result<()> {
+        self.conn.execute(
+            "INSERT INTO blocks (cid, data) VALUES (?1, ?2)",
+            (&block.cid, &block.data),
+        )?;
+        // TODO: Should we have another indicator for root blocks that isn't just the number of links?
+        // TODO: This logic should probably get pulled up and split into two parts:
+        // 1. import_block - Handles importing block into block store
+        // 2. import_block_links - Handles correctly updating links store to account for block
+        // If root block with links, then insert links
+        if !block.links.is_empty() {
+            for link_cid in &block.links {
+                let mut maybe_block_id = None;
+                if let Ok(block_id) = self.conn.query_row(
+                    "SELECT id FROM blocks b
+                    WHERE cid == (?1)",
+                    [link_cid],
+                    |row| {
+                        let id: u32 = row.get(0).unwrap();
+                        Ok(id)
+                    },
+                ) {
+                    maybe_block_id = Some(block_id);
+                }
+                self.conn.execute(
+                    "INSERT INTO links (root_cid, block_cid, block_id) VALUES(?1, ?2, ?3)",
+                    (&block.cid, link_cid, maybe_block_id),
+                )?;
+            }
+        } else {
+            // If child block w/out links...then check if the ID should be inserted into the links table
+            self.conn.execute(
+                "UPDATE links SET block_id = (SELECT id from blocks WHERE cid = ?1) WHERE block_cid = ?2",
+                (&block.cid, &block.cid),
+            )?;
+        }
+        Ok(())
+    }
+
+    fn get_available_cids(&self) -> Result<Vec<String>> {
+        let mut cids_list = Vec::<String>::new();
```

Comment:
You can probably combine all of these lines into a single statement.

Reply:
Yes! I was able to do that for most of these functions 🦀